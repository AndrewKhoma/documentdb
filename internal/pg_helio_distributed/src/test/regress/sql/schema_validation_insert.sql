SET search_path TO helio_core,helio_api,helio_api_catalog,helio_api_internal;
SET citus.next_shard_id TO 17771000;
SET helio_api.next_collection_id TO 177710;
SET helio_api.next_collection_index_id TO 177710;
set helio_api.enableSchemaValidation = true;

--------------------------------------Need $jsonSchema--------------------------------------
SELECT helio_api.create_collection_view('schema_validation_insertion', '{ "create": "col", "validator": {"$jsonSchema": {"bsonType": "object", "properties": {"a": {"bsonType": "int"}}}}, "validationLevel": "strict", "validationAction": "error"}');

SELECT helio_api.insert('schema_validation_insertion', '{"insert":"col", "documents":[{"_id":"1", "a":1}]}');
-- required not supported yet, so this should be inserted
SELECT helio_api.insert('schema_validation_insertion', '{"insert":"col", "documents":[{"_id":"2", "b":1}]}');
-- type mismatch
SELECT helio_api.insert('schema_validation_insertion','{"insert":"col", "documents":[{"_id":"3", "a":"hello"}]}');
-- batch insert
SELECT helio_api.insert('schema_validation_insertion','{"insert":"col", "documents":[{"_id":"4", "a":2},{"_id":"5", "a":3}, {"_id":"6", "a":"tt"}]}');
-- 0 documents should be inserted
SELECT shard_key_value, object_id, document from helio_api.collection('schema_validation_insertion','col') ORDER BY shard_key_value, object_id;
-- set validationAction to warn
SELECT helio_api.coll_mod('schema_validation_insertion', 'col', '{"collMod":"col", "validationAction": "warn"}');
SELECT helio_api.insert('schema_validation_insertion','{"insert":"col", "documents":[{"_id":"7", "a":"hello"}]}');
-- 1 document should be inserted
SELECT shard_key_value, object_id, document from helio_api.collection('schema_validation_insertion','col') ORDER BY shard_key_value, object_id;




---------------------------------------------Need top level operator-----------------------------------------------------
-- $expr
SELECT helio_api.create_collection_view('schema_validation_insertion', '{ "create": "col1", "validator": { "$expr": {"$eq": [ "$a", "$b" ] } } }');
SELECT helio_api.insert('schema_validation_insertion', '{"insert":"col1", "documents":[{"_id":"1", "a":1, "b":1, "c":1}]}');
SELECT helio_api.insert('schema_validation_insertion', '{"insert":"col1", "documents":[{"_id":"2", "a":3, "b":1, "c":2}]}');

-- $and
SELECT helio_api.create_collection_view('schema_validation_insertion', '{ "create": "col2", "validator": { "$and": [ { "a": { "$gt": 2 } }, {"$jsonSchema": {"bsonType": "object", "properties": {"a": {"bsonType": "int"}}}} ] } }');
SELECT helio_api.insert('schema_validation_insertion', '{"insert":"col2", "documents":[{"_id":"1", "a":3}]}');
SELECT helio_api.insert('schema_validation_insertion', '{"insert":"col2", "documents":[{"_id":"2", "a":1}]}');
set helio_api.enableBypassDocumentValidation = true;
SELECT helio_api.insert('schema_validation_insertion', '{"insert":"col2", "documents":[{"_id":"2", "a":1}], "bypassDocumentValidation": true}');
set helio_api.enableBypassDocumentValidation = false;




---------------------------------------------simple case-----------------------------------------------------
-- field 
SELECT helio_api.create_collection_view('schema_validation_insertion', '{ "create": "col3", "validator": {"a":{"$type":"int"}}}');
SELECT helio_api.insert('schema_validation_insertion', '{"insert":"col3", "documents":[{"_id":"1", "a":1}]}');
SELECT helio_api.insert('schema_validation_insertion', '{"insert":"col3", "documents":[{"_id":"2", "a":"hello"}]}');

--$merge
--todo - need to check
SELECT helio_api.insert('schema_validation_insertion','{"insert":"col_", "documents":[{"_id":"1001","a":"world"}]}');
SELECT helio_api.insert('schema_validation_insertion','{"insert":"col_", "documents":[{"_id":"1002","a":2}]}');
SELECT * FROM aggregate_cursor_first_page('schema_validation_insertion', '{ "aggregate": "col_", "pipeline": [ { "$match": { "a": { "$type": "string" }}}, {"$merge" : { "into": "col3" }} ], "cursor": { "batchSize": 1 } }', 4294967294);
SELECT * FROM aggregate_cursor_first_page('schema_validation_insertion', '{ "aggregate": "col_", "pipeline": [ { "$match": { "a": { "$type": "int" }}}, {"$merge" : { "into": "col3" }} ], "cursor": { "batchSize": 1 } }', 4294967294);
SELECT shard_key_value, object_id, document from helio_api.collection('schema_validation_insertion','col3') ORDER BY shard_key_value, object_id;

-- sharded collection test
SELECT helio_api.shard_collection('schema_validation_insertion', 'col3', '{ "a": "hashed" }', false);
SELECT helio_api.insert('schema_validation_insertion', '{"insert":"col3", "documents":[{"_id":"1", "a":"hello"}]}');
SELECT helio_api.insert('schema_validation_insertion', '{"insert":"col3", "documents":[{"_id":"2", "a":5}]}');
SELECT helio_api.insert('schema_validation_insertion', '{"insert":"col3", "documents":[{"_id":"3", "a":2}, {"_id":"4", "a":3}, {"_id":"5", "a":4}, {"_id":"6", "a":"string"}]}');
-- 5 documents should be inserted
SELECT shard_key_value, object_id, document from helio_api.collection('schema_validation_insertion','col3') ORDER BY shard_key_value, object_id;
-- set validationAction to warn
SELECT helio_api.coll_mod('schema_validation_insertion', 'col3', '{"collMod":"col3", "validationAction": "warn"}');
SELECT helio_api.insert('schema_validation_insertion','{"insert":"col3", "documents":[{"_id":"7", "a":"hello"}]}');
-- 6 document should be inserted
SELECT shard_key_value, object_id, document from helio_api.collection('schema_validation_insertion','col3') ORDER BY shard_key_value, object_id;
