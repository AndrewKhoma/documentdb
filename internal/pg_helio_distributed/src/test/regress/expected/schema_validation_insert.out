SET search_path TO helio_core,helio_api,helio_api_catalog,helio_api_internal;
SET citus.next_shard_id TO 17771000;
SET helio_api.next_collection_id TO 177710;
SET helio_api.next_collection_index_id TO 177710;
set helio_api.enableSchemaValidation = true;
--------------------------------------Need $jsonSchema--------------------------------------
SELECT helio_api.create_collection_view('schema_validation_insertion', '{ "create": "col", "validator": {"$jsonSchema": {"bsonType": "object", "properties": {"a": {"bsonType": "int"}}}}, "validationLevel": "strict", "validationAction": "error"}');
NOTICE:  creating collection
         create_collection_view         
---------------------------------------------------------------------
 { "ok" : { "$numberDouble" : "1.0" } }
(1 row)

SELECT helio_api.insert('schema_validation_insertion', '{"insert":"col", "documents":[{"_id":"1", "a":1}]}');
ERROR:  unknown operator: $jsonSchema
-- required not supported yet, so this should be inserted
SELECT helio_api.insert('schema_validation_insertion', '{"insert":"col", "documents":[{"_id":"2", "b":1}]}');
ERROR:  unknown operator: $jsonSchema
-- type mismatch
SELECT helio_api.insert('schema_validation_insertion','{"insert":"col", "documents":[{"_id":"3", "a":"hello"}]}');
ERROR:  unknown operator: $jsonSchema
-- batch insert
SELECT helio_api.insert('schema_validation_insertion','{"insert":"col", "documents":[{"_id":"4", "a":2},{"_id":"5", "a":3}, {"_id":"6", "a":"tt"}]}');
ERROR:  unknown operator: $jsonSchema
-- 0 documents should be inserted
SELECT shard_key_value, object_id, document from helio_api.collection('schema_validation_insertion','col') ORDER BY shard_key_value, object_id;
 shard_key_value | object_id | document 
---------------------------------------------------------------------
(0 rows)

-- set validationAction to warn
SELECT helio_api.coll_mod('schema_validation_insertion', 'col', '{"collMod":"col", "validationAction": "warn"}');
             coll_mod              
---------------------------------------------------------------------
 { "ok" : { "$numberInt" : "1" } }
(1 row)

SELECT helio_api.insert('schema_validation_insertion','{"insert":"col", "documents":[{"_id":"7", "a":"hello"}]}');
                                         insert                                         
---------------------------------------------------------------------
 ("{ ""n"" : { ""$numberInt"" : ""1"" }, ""ok"" : { ""$numberDouble"" : ""1.0"" } }",t)
(1 row)

-- 1 document should be inserted
SELECT shard_key_value, object_id, document from helio_api.collection('schema_validation_insertion','col') ORDER BY shard_key_value, object_id;
 shard_key_value |  object_id   |            document            
---------------------------------------------------------------------
          177711 | { "" : "7" } | { "_id" : "7", "a" : "hello" }
(1 row)

---------------------------------------------Need top level operator-----------------------------------------------------
-- $expr
SELECT helio_api.create_collection_view('schema_validation_insertion', '{ "create": "col1", "validator": { "$expr": {"$eq": [ "$a", "$b" ] } } }');
NOTICE:  creating collection
         create_collection_view         
---------------------------------------------------------------------
 { "ok" : { "$numberDouble" : "1.0" } }
(1 row)

SELECT helio_api.insert('schema_validation_insertion', '{"insert":"col1", "documents":[{"_id":"1", "a":1, "b":1, "c":1}]}');
ERROR:  $expr can only be applied to the top-level document
SELECT helio_api.insert('schema_validation_insertion', '{"insert":"col1", "documents":[{"_id":"2", "a":3, "b":1, "c":2}]}');
ERROR:  $expr can only be applied to the top-level document
-- $and
SELECT helio_api.create_collection_view('schema_validation_insertion', '{ "create": "col2", "validator": { "$and": [ { "a": { "$gt": 2 } }, {"$jsonSchema": {"bsonType": "object", "properties": {"a": {"bsonType": "int"}}}} ] } }');
NOTICE:  creating collection
         create_collection_view         
---------------------------------------------------------------------
 { "ok" : { "$numberDouble" : "1.0" } }
(1 row)

SELECT helio_api.insert('schema_validation_insertion', '{"insert":"col2", "documents":[{"_id":"1", "a":3}]}');
ERROR:  unknown top level operator: $jsonSchema. If you have a field name that starts with a '$' symbol, consider using $getField or $setField.
SELECT helio_api.insert('schema_validation_insertion', '{"insert":"col2", "documents":[{"_id":"2", "a":1}]}');
ERROR:  unknown top level operator: $jsonSchema. If you have a field name that starts with a '$' symbol, consider using $getField or $setField.
set helio_api.enableBypassDocumentValidation = true;
SELECT helio_api.insert('schema_validation_insertion', '{"insert":"col2", "documents":[{"_id":"2", "a":1}], "bypassDocumentValidation": true}');
                                         insert                                         
---------------------------------------------------------------------
 ("{ ""n"" : { ""$numberInt"" : ""1"" }, ""ok"" : { ""$numberDouble"" : ""1.0"" } }",t)
(1 row)

set helio_api.enableBypassDocumentValidation = false;
---------------------------------------------simple case-----------------------------------------------------
-- field 
SELECT helio_api.create_collection_view('schema_validation_insertion', '{ "create": "col3", "validator": {"a":{"$type":"int"}}}');
NOTICE:  creating collection
         create_collection_view         
---------------------------------------------------------------------
 { "ok" : { "$numberDouble" : "1.0" } }
(1 row)

SELECT helio_api.insert('schema_validation_insertion', '{"insert":"col3", "documents":[{"_id":"1", "a":1}]}');
                                         insert                                         
---------------------------------------------------------------------
 ("{ ""n"" : { ""$numberInt"" : ""1"" }, ""ok"" : { ""$numberDouble"" : ""1.0"" } }",t)
(1 row)

SELECT helio_api.insert('schema_validation_insertion', '{"insert":"col3", "documents":[{"_id":"2", "a":"hello"}]}');
                                                                                                                        insert                                                                                                                        
---------------------------------------------------------------------
 ("{ ""n"" : { ""$numberInt"" : ""0"" }, ""ok"" : { ""$numberDouble"" : ""1.0"" }, ""writeErrors"" : [ { ""index"" : { ""$numberInt"" : ""0"" }, ""code"" : { ""$numberInt"" : ""525074461"" }, ""errmsg"" : ""Document failed validation"" } ] }",f)
(1 row)

--$merge
--todo - need to check
SELECT helio_api.insert('schema_validation_insertion','{"insert":"col_", "documents":[{"_id":"1001","a":"world"}]}');
NOTICE:  creating collection
                                         insert                                         
---------------------------------------------------------------------
 ("{ ""n"" : { ""$numberInt"" : ""1"" }, ""ok"" : { ""$numberDouble"" : ""1.0"" } }",t)
(1 row)

SELECT helio_api.insert('schema_validation_insertion','{"insert":"col_", "documents":[{"_id":"1002","a":2}]}');
                                         insert                                         
---------------------------------------------------------------------
 ("{ ""n"" : { ""$numberInt"" : ""1"" }, ""ok"" : { ""$numberDouble"" : ""1.0"" } }",t)
(1 row)

SELECT * FROM aggregate_cursor_first_page('schema_validation_insertion', '{ "aggregate": "col_", "pipeline": [ { "$match": { "a": { "$type": "string" }}}, {"$merge" : { "into": "col3" }} ], "cursor": { "batchSize": 1 } }', 4294967294);
                                                                      cursorpage                                                                      | continuation | persistconnection | cursorid 
---------------------------------------------------------------------
 { "cursor" : { "id" : { "$numberLong" : "0"}, "ns" : "schema_validation_insertion.col_", "firstBatch" : [  ] }, "ok" : { "$numberDouble" : "1.0" } } |              | f                 |        0
(1 row)

SELECT * FROM aggregate_cursor_first_page('schema_validation_insertion', '{ "aggregate": "col_", "pipeline": [ { "$match": { "a": { "$type": "int" }}}, {"$merge" : { "into": "col3" }} ], "cursor": { "batchSize": 1 } }', 4294967294);
                                                                      cursorpage                                                                      | continuation | persistconnection | cursorid 
---------------------------------------------------------------------
 { "cursor" : { "id" : { "$numberLong" : "0"}, "ns" : "schema_validation_insertion.col_", "firstBatch" : [  ] }, "ok" : { "$numberDouble" : "1.0" } } |              | f                 |        0
(1 row)

SELECT shard_key_value, object_id, document from helio_api.collection('schema_validation_insertion','col3') ORDER BY shard_key_value, object_id;
 shard_key_value |    object_id    |                     document                     
---------------------------------------------------------------------
          177714 | { "" : "1" }    | { "_id" : "1", "a" : { "$numberInt" : "1" } }
          177714 | { "" : "1001" } | { "_id" : "1001", "a" : "world" }
          177714 | { "" : "1002" } | { "_id" : "1002", "a" : { "$numberInt" : "2" } }
(3 rows)

-- sharded collection test
SELECT helio_api.shard_collection('schema_validation_insertion', 'col3', '{ "a": "hashed" }', false);
 shard_collection 
---------------------------------------------------------------------
 
(1 row)

SELECT helio_api.insert('schema_validation_insertion', '{"insert":"col3", "documents":[{"_id":"1", "a":"hello"}]}');
                                                                                                                        insert                                                                                                                        
---------------------------------------------------------------------
 ("{ ""n"" : { ""$numberInt"" : ""0"" }, ""ok"" : { ""$numberDouble"" : ""1.0"" }, ""writeErrors"" : [ { ""index"" : { ""$numberInt"" : ""0"" }, ""code"" : { ""$numberInt"" : ""525074461"" }, ""errmsg"" : ""Document failed validation"" } ] }",f)
(1 row)

SELECT helio_api.insert('schema_validation_insertion', '{"insert":"col3", "documents":[{"_id":"2", "a":5}]}');
                                         insert                                         
---------------------------------------------------------------------
 ("{ ""n"" : { ""$numberInt"" : ""1"" }, ""ok"" : { ""$numberDouble"" : ""1.0"" } }",t)
(1 row)

SELECT helio_api.insert('schema_validation_insertion', '{"insert":"col3", "documents":[{"_id":"3", "a":2}, {"_id":"4", "a":3}, {"_id":"5", "a":4}, {"_id":"6", "a":"string"}]}');
                                                                                                                        insert                                                                                                                        
---------------------------------------------------------------------
 ("{ ""n"" : { ""$numberInt"" : ""3"" }, ""ok"" : { ""$numberDouble"" : ""1.0"" }, ""writeErrors"" : [ { ""index"" : { ""$numberInt"" : ""3"" }, ""code"" : { ""$numberInt"" : ""525074461"" }, ""errmsg"" : ""Document failed validation"" } ] }",f)
(1 row)

-- 5 documents should be inserted
SELECT shard_key_value, object_id, document from helio_api.collection('schema_validation_insertion','col3') ORDER BY shard_key_value, object_id;
   shard_key_value    |    object_id    |                     document                     
---------------------------------------------------------------------
 -4918719581749358852 | { "" : "4" }    | { "_id" : "4", "a" : { "$numberInt" : "3" } }
 -1389566185330078543 | { "" : "1002" } | { "_id" : "1002", "a" : { "$numberInt" : "2" } }
 -1389566185330078543 | { "" : "3" }    | { "_id" : "3", "a" : { "$numberInt" : "2" } }
  1786987034919379147 | { "" : "2" }    | { "_id" : "2", "a" : { "$numberInt" : "5" } }
  4004935074940511305 | { "" : "5" }    | { "_id" : "5", "a" : { "$numberInt" : "4" } }
  4322365043291501017 | { "" : "1" }    | { "_id" : "1", "a" : { "$numberInt" : "1" } }
  5505078723908235407 | { "" : "1001" } | { "_id" : "1001", "a" : "world" }
(7 rows)

-- set validationAction to warn
SELECT helio_api.coll_mod('schema_validation_insertion', 'col3', '{"collMod":"col3", "validationAction": "warn"}');
             coll_mod              
---------------------------------------------------------------------
 { "ok" : { "$numberInt" : "1" } }
(1 row)

SELECT helio_api.insert('schema_validation_insertion','{"insert":"col3", "documents":[{"_id":"7", "a":"hello"}]}');
                                         insert                                         
---------------------------------------------------------------------
 ("{ ""n"" : { ""$numberInt"" : ""1"" }, ""ok"" : { ""$numberDouble"" : ""1.0"" } }",t)
(1 row)

-- 6 document should be inserted
SELECT shard_key_value, object_id, document from helio_api.collection('schema_validation_insertion','col3') ORDER BY shard_key_value, object_id;
   shard_key_value    |    object_id    |                     document                     
---------------------------------------------------------------------
 -4918719581749358852 | { "" : "4" }    | { "_id" : "4", "a" : { "$numberInt" : "3" } }
 -1389566185330078543 | { "" : "1002" } | { "_id" : "1002", "a" : { "$numberInt" : "2" } }
 -1389566185330078543 | { "" : "3" }    | { "_id" : "3", "a" : { "$numberInt" : "2" } }
  1786987034919379147 | { "" : "2" }    | { "_id" : "2", "a" : { "$numberInt" : "5" } }
  3587377943449965360 | { "" : "7" }    | { "_id" : "7", "a" : "hello" }
  4004935074940511305 | { "" : "5" }    | { "_id" : "5", "a" : { "$numberInt" : "4" } }
  4322365043291501017 | { "" : "1" }    | { "_id" : "1", "a" : { "$numberInt" : "1" } }
  5505078723908235407 | { "" : "1001" } | { "_id" : "1001", "a" : "world" }
(8 rows)

